Add Double Click to Open in Default Editor

digitalmaster opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #2620

## Description

- [x] Add Double Click To Open Changed file In Default Program
- [x] Add same interaction to all changes list views
- [x] [REVERTED] ~~Add same interaction to double click diff line number~~
- [x] Investigate: Bold "Open with Default Program" context menu item (Windows)

### Preview

https://user-images.githubusercontent.com/1279335/233219332-a7469a62-007e-4eee-a720-19ff636c2489.mp4

# Testing Instructions

**Double Clicking Files Lists**
1. Find a change list (Main Change List, History Change List, Stash Change List)
2. Double Click on the file name

**Expectation:** Should open file in default editor

**Double Clicking Diff Line Numbers**
1. Find a diff view  (Main Change List, History Change List, Stash Change List)
2. Double Click on line number

**Expectation:** If vscode `code` command is installed on path, should open file at line number in default editor.

https://user-images.githubusercontent.com/1279335/236348957-315e6305-d376-4102-a48f-92c4e12c0d40.mp4

**Expectation:** Otherwise, it shows this dialog message
![CleanShot 2023-05-04 at 15 58 39@2x](https://user-images.githubusercontent.com/1279335/236347763-0d5fa8ab-ad68-44a9-a18d-dfcc9bc8f824.png)
**Expectation:** If the default editor doesn't have a supported `editorOpenToLineCommands` function defined, it just opens the file without going to that specific line. 

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes:

digitalmaster committed:
Add Open on DoubleClick Interaction

tidy-dev reviewed (dismissed):
👋 Great start! This approach does work. However, it is not consistent with our current approach to interacting with list items. If you look in the `changes-list.tsx` at the `<List>` element. You will see an implementation of `onRowClick` and `onRowContextMenu`. To be consistent, it would be better to implement a `onRowDoubleClick`. This would also allow you to reuse some of the logic in the `getDefaultContextMenu` and `getOpenInExternalEditorMenuItem` for opening in an external editor. (May even be ideal to create a shared method for the action part of the context menu bit so both implementations are ensured to do the same thing)

digitalmaster committed:
PR Feedback: Refactor to use `onRowDoubleClick` handler

Added handler in parent `sidebar.tsx` since that seems to be where all the primary handlers are defined (ie. `onRowClick` and `onFileSelectionChanged`).

digitalmaster commented:
Hi @tidy-dev 👋🏽 - Thanks for taking the time to review. 

Great feedback... Iterating now...

digitalmaster committed:
PR Feedback: Create shared `onOpenItemInExternalEditor` handler`

digitalmaster committed:
Wire up history view

digitalmaster committed:
Add to changes diff line number

digitalmaster committed:
Wire up commit history changes diff line numbers

digitalmaster committed:
Wire up remaining views

digitalmaster commented:
Had some time on a plane to work on this yesterday. Would love some more feedback. Also anything I may be missing to get this ready to merge.

> Re: Bolding default option for windows

Hmm... [doesn't look](https://github.com/electron/electron/issues/3908) like this is currently possible in electron.

digitalmaster commented on app/src/lib/editors/launch.ts:
```diff
@@ -32,10 +35,45 @@ export async function launchExternalEditor(
+    const execName = 'code'
+    if (!(await executableExists(execName))) {
+      throw new Error(
+        `Could not find the '${execName}' executable for '${editor.editor}' on path. Please install and try again.`
```
NIT: Maybe add link to instructions for installing the vscode cli?

digitalmaster commented on app/src/lib/editors/launch.ts:
```diff
@@ -32,10 +35,45 @@ export async function launchExternalEditor(
   if (editor.usesShell) {
     spawn(`"${editorPath}"`, [`"${fullPath}"`], { ...opts, shell: true })
   } else if (__DARWIN__) {
```
NIT: Because of this guard diff double clicks will not go to the specified line.

digitalmaster reviewed (commented):
Self review I.

sergiou87 reviewed (changes requested):
Thanks for your contribution! 💖 

I tried this out and noticed that double-clicking files from the list of files changed inside of a commit doesn't work as intended.

The scenario I have is this:
1. Open VS Code with the `desktop/desktop` repo
2. Open GitHub Desktop with a different repo selected (let's say `sergiou87/test-repo`), a repo with `README.md`
3. Look for a commit that changed `README.md`
4. Double click on `README.md` inside of such commit

Instead of opening `README.md` of `sergiou87/test-repo`, it opens `README.md` in the current instance of VS Code with `desktop/desktop` 🤔 

Aside from that bug, I just noticed you also added "double-click on diff line to open file at that line", which I _think_ is something we didn't agree on? I will check with the team next week and get back to you 😄

digitalmaster commented:
Thanks for taking a look 🙌🏽

> Instead of opening README.md of sergiou87/test-repo, it opens README.md in the current instance of VS Code with desktop/desktop 🤔

Excellent find!! 🥳 I was able to reproduce it. Looks like I need to go through all the code paths and ensure that we're always generating a full path instead of a relative path everywhere. 

Gonna also see if I can refactor this so that this happens consistently in one place.

> Aside from that bug, I just noticed you also added "double-click on diff line to open file at that line", which I think is something we didn't agree on? I will check with the team next week and get back to you 😄

Oh I see. My apologies. I'm not familiar with what specifically was agreed upon or how things like this work on your team but boy do I hope this is allowed. To be fair tho the only place this was mentioned was [here](https://github.com/desktop/desktop/issues/2620#issuecomment-1318914036):
![CleanShot 2023-05-05 at 11 53 29@2x](https://user-images.githubusercontent.com/1279335/236544239-06a4ed15-bb86-4511-8809-cd10dbb2bbfd.png)


Any reason you think the team might ok with double click to a file but not to the specific line number? I've been using it while developing this feature and I really think users will love this. It's so common that i'd wanna make a change to a specific line while reviewing a diff.. this makes that so fast. Please please please 🤞🏽🙏🏽 😭 😅

Thanks again for taking the time to review <3

digitalmaster committed:
Refactor files list to use shared `onOpenInExternalEditor` fn

This refactor ensures that all double click event handlers bubble up to the same `onOpenInExternalEditor` fn in app.tsx.

This is where we prefix the repository path to the relative path which fixes a bug we caught where wrong files would open the opened project had a file at the same relative path (ie. README.md).

sergiou87 commented:
I'm sorry but we just chatted and are not convinced about it 😕 As mentioned: it's not very discoverable and kind of conflicts with the existing interactions. Adding something like that would require more thought and discussion, but our current priorities don't give us much time for it 😞 

We think it's better to focus on "double-click on file" in this PR, and anything beyond that should be discussed that in a new issue. But again, we don't have the bandwidth to engage in such discussion right now, so it might take a while until we take a decision 😓

digitalmaster committed:
Remove "Go to specific line" feature

digitalmaster commented:
Got it.. i'll open a new issue. I just finished reverting the feature. Take a look when you have a moment 🙏🏽

sergiou87 commented on app/src/ui/changes/changes-list.tsx:
```diff
@@ -151,6 +151,14 @@ interface IChangesListProps {
   readonly onOpenItem: (path: string) => void
+
+  /**
+   * Called to open a file it its the default external editor
```
Nit:
```suggestion
   * Called to open a file it the default external editor
```

sergiou87 commented on app/src/ui/history/selected-commits.tsx:
```diff
@@ -116,6 +116,17 @@ export class SelectedCommits extends React.Component<
+    const files = this.props.changesetData.files
+    const file = files[row]
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
This should be either `log.error(…)` or nothing. I think we can just remove it:
```suggestion
```

digitalmaster replied:
Yea, agreed this is unnecessary (was just blindly following a pattern i'd seen 🙈) 👍🏽

sergiou87 commented on app/src/ui/open-pull-request/pull-request-files-changed.tsx:
```diff
@@ -224,6 +231,17 @@ export class PullRequestFilesChanged extends React.Component<
+    const files = this.props.files
+    const file = files[row]
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Same:
```suggestion
```

sergiou87 commented on app/src/ui/stashing/stash-diff-viewer.tsx:
```diff
@@ -64,11 +71,26 @@ export class StashDiffViewer extends React.PureComponent<IStashDiffViewerProps>
+    const files = this.getFiles()
+    const file = files[row]
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Same:
```suggestion
```

sergiou87 commented on app/src/ui/changes/sidebar.tsx:
```diff
@@ -249,6 +249,26 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
+    const file = workingDirectory.files[row]
+
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Same:
```suggestion
```

sergiou87 commented on app/src/ui/changes/sidebar.tsx:
```diff
@@ -249,6 +249,26 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
     openFile(fullPath, this.props.dispatcher)
   }
+  /**
+   * Called to open a file it its the default external editor
```
Nit:
```suggestion
   * Called to open a file it the default external editor
```

sergiou87 commented on app/src/ui/changes/changes-list.tsx:
```diff
@@ -981,6 +982,7 @@ export class ChangesList extends React.Component<
               isCommitting: isCommitting,
             }}
             onRowClick={this.props.onRowClick}
+            onRowDoubleClick={this.props.onRowDoubleClick}
```
Just curious: why bubbling this up instead of also calling `onOpenItemInExternalEditor`?

digitalmaster replied:
Ah good catch 🥳 -  no need to hoist this any further 👍🏽

sergiou87 reviewed (changes requested):
It looks good overall! Thank you for all those changes 🙇‍♂️ 
I found a few nits/typos, some `console.blahblah` that should be removed, but other than that looks good. 

I also wonder why in some controls you have bubbled up both double click and "open in external editor" instead of grouping both in the latter 🤔

digitalmaster committed:
PR Feedback: Typo

digitalmaster committed:
PR Feedback: Remove unnecessary error handling

digitalmaster committed:
PR Feedback: Stop `onRowDoubleClick` hoisting @ changes-list

digitalmaster commented:
Thanks for the feedback. Iterated. Give it another pass when you have a chance 🙏🏽

sergiou87 commented on app/src/ui/history/selected-commits.tsx:
Line 400 needs to change from
```typescript
action: () => this.props.onOpenInExternalEditor(fullPath),
```
to
```typescript
action: () => this.props.onOpenInExternalEditor(file.path),
```

digitalmaster replied:
Oh that's right... Nice catch! Thanks for testing 🙌🏽

sergiou87 commented on app/src/ui/changes/sidebar.tsx:
```diff
@@ -249,6 +249,14 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
     openFile(fullPath, this.props.dispatcher)
   }
+  /**
+   * Called to open a file in its the default external editor
```
I noticed you left part of the typo here:
```suggestion
   * Called to open a file in the default external editor
```

digitalmaster replied:
ha! You're patience knows no bounds.. I'll get it right eventually 😆 🤦🏽‍♂️

sergiou87 replied:
LOL no worries 🫂 I make those mistakes (and much worse) all the time 😂

sergiou87 reviewed (changes requested):
Thank you for those quick changes! 💖 

Unfortunately, I tested it and noticed that going to a commit -> right click on a file -> Open in _Whatever_ doesn't work anymore, because clicking on that still uses the full path instead of the relative one.

I think that should be it 🤞

digitalmaster committed:
Bug Fix: "Right click to open" argument path should be relative

digitalmaster committed:
PR Feedback: Fix typo round II

sergiou87 reviewed (approved):
Code looks great and works like a charm! Thank you for contribution and your patience through all this process 🙇‍♂️🙇‍♂️🙇‍♂️🙇‍♂️

digitalmaster commented:
It was my pleasure. Thank you for the time to review and get this ready for primetime 🙌🏽

mr21 commented:
Hi great feature :) thanks you for that

But currently this feature works also when we double click on the checkbox:
![image](https://github.com/desktop/desktop/assets/850754/472b7b87-091c-424b-ab98-180d57342df6)

And we can find a reason to double click on those checkbox to reset the particular line by line selection

sergiou87 commented:
Hey @mr21 👋 

Thanks for reporting that, I missed that comment and just ran into it 😱 We will take a look at it. It's not something I've suffered myself but I understand how it could be super annoying.

**EDIT:** Fixed in #17229

mr21 commented:
Cool, it's working perfectly now 🥳

digitalmaster commented:
This was removed in the latest version. Are there any plans to add this back?

tidy-dev commented:
@digitalmaster What version are seeing it not working? Just tried on the latest beta and it worked as expected.

digitalmaster commented:
Oh sorry i'm talking about click to go directly to specific lines. I forgot that you guys didn't think it was worth adding that part (reverted and made [separate issue](https://github.com/desktop/desktop/issues/20254)) so I kept my own build with the feature and have been using it for years now.

That said, are you sure you guys didn't add support for this afterwards? IIRC at some point I stopped having to rollback  to my personal build after every update?

tidy-dev commented:
Ah. I could be wrong, but I do not believe we added support for that.

digitalmaster commented:
Ok good to know there's a small chance i'm not completely making that up 😆.

For me, this is still the most use action when i'm reviewing a diff - I'm going through my changes and I see a line that want to tweak, I double click to go right to it (no need to go to file then scroll or search).

digitalmaster commented:
@tidy-dev Would the team open to a PR that adds this feature back? If so I could make some time to do this again in the coming weeks.

MarkHaakman commented:
@digitalmaster If the main GitHub team still does not want to add support for double clicking lines to open it in an editor (which is in their right of course), you could also open a pull request to https://github.com/pol-rivero/github-desktop-plus , which is an actively maintained fork with additional features. 

Getting this functionality in the main repo is of course far nicer, but if the team decides against it, this is the next best alternative I guess.
